### PR TITLE
Fix flaky test and add more test logging

### LINKED ISSUE #6354: Flaky `DashboardServiceTests.WatchResources_ResourceHasCommands_CommandsSentWithResponse`
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=845546
Build error leg or test failing: Aspire.Hosting.Tests.Dashboard.DashboardServiceTests.WatchResources_ResourceHasCommands_CommandsSentWithResponse
Pull request: https://github.com/dotnet/aspire/pull/6235

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": "Assert.Single() Failure: The collection was empty",
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=845546
**Error message validated:** `[Assert.Single() Failure: The collection was empty`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 10/17/2024 7:29:21 AM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[847136](https://dev.azure.com/dnceng-public/public/_build/results?buildId=847136)|dotnet/aspire|[Aspire.Hosting.Tests.Dashboard.DashboardServiceTests.WatchResources_ResourceHasCommands_CommandsSentWithResponse](https://dev.azure.com/dnceng-public/public/_build/results?buildId=847136&view=ms.vss-test-web.build-test-results-tab&runId=21987022&resultId=100210)|dotnet/aspire#6362|
|[847076](https://dev.azure.com/dnceng-public/public/_build/results?buildId=847076)|dotnet/aspire|[Aspire.Hosting.Tests.Dashboard.DashboardServiceTests.WatchResources_ResourceHasCommands_CommandsSentWithResponse](https://dev.azure.com/dnceng-public/public/_build/results?buildId=847076&view=ms.vss-test-web.build-test-results-tab&runId=21984962&resultId=100049)|dotnet/aspire#6275|
|[845546](https://dev.azure.com/dnceng-public/public/_build/results?buildId=845546)|dotnet/aspire|[Aspire.Hosting.Tests.Dashboard.DashboardServiceTests.WatchResources_ResourceHasCommands_CommandsSentWithResponse](https://dev.azure.com/dnceng-public/public/_build/results?buildId=845546&view=ms.vss-test-web.build-test-results-tab&runId=21938060&resultId=100125)|dotnet/aspire#6235|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|2|3|3|

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

Adds more logging to tests.
Also improve flaky test to avoid potential race condition.

Fixes https://github.com/dotnet/aspire/issues/6354

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6366)

### CODE REVIEW

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs [hunk 1]
```diff
@@ -93,29 +102,37 @@ public async Task WatchResources_ResourceHasCommands_CommandsSentWithResponse()
...
             isHighlighted: true);
 
+        logger.LogInformation("Publishing resource.");
         await resourceNotificationService.PublishUpdateAsync(testResource, s =>
         {
             return s with { State = new ResourceStateSnapshot("Starting", null) };
         });
-        await resourceNotificationService.WaitForResourceAsync(testResource.Name, r =>
+
+        logger.LogInformation("Waiting for the resource with a command. Required so reource is always in the service's initial data collection");
```
Comment (JamesNK) [inline tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:111]:
```suggestion
        logger.LogInformation("Waiting for the resource with a command. Required so added resource is always in the service's initial data collection");
```

**Review by mitchdenny — APPROVED** [review]

**Comment by adamint** [comment]
Test failure seemed unrelated, re-running.

File: tests/Shared/TaskExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return task.AsTask().TimeoutAfter(TimeSpan.FromMilliseconds(milliseconds), filePath, lineNumber);
+    }
+
+    public static Task<T> DefaultTimeout<T>(this ValueTask<T> task, TimeSpan timeout, [CallerFilePath] string? filePath = null, [CallerLineNumber] int lineNumber = default)
+    {
+        return task.AsTask().TimeoutAfter(timeout, filePath, lineNumber);
+    }
+
+    [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
```
Comment (eerhardt) [inline tests/Shared/TaskExtensions.cs:62]: 1. Do we have public analyzers turned on for our tests? That seems wrong.
2. What does `"Required to maintain compatibility"` mean for our tests? I don't think we need to worry about compatibility.

Comment (JamesNK) [reply tests/Shared/TaskExtensions.cs:62]: It’s copied from dotnet/aspnetcore

Comment (eerhardt) [reply tests/Shared/TaskExtensions.cs:62]: It would be good to add that in a comment on the file

**Pull request merged**